### ISSUE TITLE
MMR leaves in off-chain storage can be overwritten by forks

### ISSUE THREAD

**Issue by tomusdrw** [paritytech/grandpa-bridge-gadget#336]
Currently, the leaves in Offchain DB are keyed only with their position ([see code](https://github.com/paritytech/substrate/blob/5fa8dc0f126a4e3c9d5bf02db0fd4206f953403d/frame/merkle-mountain-range/src/lib.rs#L231-L233)). The premise was, that since the leaf content will be the same on every fork, it's fine if the Indexing API overwrites them.

However, this is incorrect for forks that are couple of blocks deeper, since the leaves may contain a different `parent_hash`. This can cause issues with off-chain proof verification or generation. I.e. importing some forks can cause offchain data to be overwritten.

The solutions I see:
1. Add blockhash (parenthash) to the offchain indexing key - this complicates the logic of the custom RPC / node retrieval, but doesn't require any changes to Indexing API.
2. Change Indexing API to only store data from canonical chain - this is way more complicated, but feels more correct. It's not enough to not write to storage on non-canonical blocks, cause they might become canonical at some point, so we would need to maintain some DB overlay and make sure the forks are handled correctly (see also "fork-aware local storage for offchain workers")
3. Support secondary key in `Indexing API`. Basically under `primary_key` we would store a `Vec<(secondary_key, value)>` - by default we could even make the secondary key a block hash (i.e. non-configurable from runtime perspective), and the off-chain logic would simply dig a correct value. The advantage of this solution over (1) is that the retrieval should be a bit simpler and we would only need to do extra verification logic for actual forks (i.e. pick the `value` that comes from canonical `secondary_key`).

I'm leaning more towards a general solution for `Indexing API` rather than having something MMR specific, since all other usages of Indexing API will run into the exact same issue.

**Comment by Wizdave97** [comment]
@tomusdrw  concerning this, I'm working on the third suggestion, if it's implemented for the indexing API does it also need to be implemented for  `fn local_storage_set(&mut self, kind: StorageKind, key: &[u8], value: &[u8]) {}` in `DbExternalities` also?

**Comment by Wizdave97** [comment]
@bkchr  my current implementation is here https://github.com/paritytech/substrate/pull/10654 and I'm implementing the third suggestion in the issue description

But the problem I am facing now is that when the value for this primary key is retrieved from the offchaindb by an offchain worker, it will be a Vec<(secondary_key, value)> and it becomes difficult to know which of the values in this vector is from the canonical chain

So I'm looking for a way to remove all values from forks from this vector before the changes to the overlay are written to the actual database, the secondary_key is meant to be a block_hash for this purpose. 

I need some insight on how to do that

**Comment by bkchr** [comment]
Either you start storing `block_hash => parent_block_hash` in the offchain db yourself, so you can do this mapping or offchain db is made fork aware.

**Comment by tomusdrw** [comment]
> Either you start storing block_hash => parent_block_hash in the offchain db yourself, so you can do this mapping or offchain db is made fork aware.

Not really, we don't need to store any mapping of this kind in offchain DB, we already have `parent_block_hash` in the runtime. Making offchain db fork-aware is solution (2), which is an alternative way to implement that.

> and it becomes difficult to know which of the values in this vector is from the canonical chain

Yes, well that's the clue of that issue. There are two ways of approaching that, which I see:
1. Either you use `block_hash` (i.e. current block hash, not available in the runtime, it has to be added before writing to DB, but after it's actually known) as a secondary key and from offchain DB API perspective you don't really have a `Vec<>` but rather simply single value, cause the underlying getter API is selecting the right element from the vector, so it's completely transparent,
2. Or you expose the entire list to the offchain worker and the offchain worker can pick the value it wants. i.e. selected the `secondary_key` it's interested in. We can have offchain workers try a bunch of different combinations of values to see which one is canon (there shouldn't be that many since forks should be shallow and rare), or we may somehow think of a some kind of chain/runtime-specific secondary key that will help us figure out the right fork (I don't think we can come up with something generic here).

**Comment by acatangiu** [comment]
Opened https://github.com/paritytech/polkadot-sdk/issues/1125 to track a generic solution in substrate.

**Comment by acatangiu** [comment]
Fixed in https://github.com/paritytech/substrate/pull/11594
